vendors/infineon/secure_elements/pkcs11/iot_pkcs11_trustx.c

```diff
@@ -0,0 +1,3298 @@
+    uint8_t * pcLabel = NULL;
+
+    /* Avoid warnings about unused parameters. */
+    ( void ) xSession;
+
+    if( (CKR_OK != xResult) || ( NULL == pxTemplate ) || ( 0 == ulCount ) )
+    {
+        xResult = CKR_ARGUMENTS_BAD;
+    }
+    else
+    {
+        /*
+         * Copy the object into a buffer.
+         */
+
+        prvFindObjectInListByHandle( xObject, &xPalHandle, &pcLabel, &xSize ); /*pcLabel and xSize are ignored. */
+
+        if( xPalHandle != CK_INVALID_HANDLE && xPalHandle != eAwsDevicePrivateKey && xPalHandle != eAwsTestPrivateKey)
+        {
+            xResult = PKCS11_PAL_GetObjectValue( xPalHandle, &pxObjectValue, &ulLength, &xIsPrivate );
+        }
+        else if (xPalHandle == CK_INVALID_HANDLE)
+        {
+            xResult = CKR_DATA_INVALID;
+        }
+    }
+
+    if ( CKR_OK != xResult)
+    {
+        xResult = CKR_DATA_INVALID;
+    }
+    else {
+        switch (xPalHandle) {
+        case eAwsDevicePrivateKey:
+        case eAwsTestPrivateKey:
+            xClass = CKO_PRIVATE_KEY;
+            break;
+        case eAwsDevicePublicKey:
+        case eAwsTestPublicKey:
+            xClass = CKO_PUBLIC_KEY;
+            break;
+        case eAwsDeviceCertificate:
+        case eAwsTestCertificate:
+            xClass = CKO_CERTIFICATE;
+            break;
+        default:
+            xResult = CKR_DATA_INVALID;
+            break;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        for( iAttrib = 0; iAttrib < ulCount && CKR_OK == xResult; iAttrib++ )
+        {
+            switch( pxTemplate[ iAttrib ].type )
+            {
+                case CKA_CLASS:
+
+                    if( pxTemplate[ iAttrib ].pValue == NULL )
+                    {
+                        pxTemplate[ iAttrib ].ulValueLen = sizeof( CK_OBJECT_CLASS );
+                    }
+                    else
+                    {
+                        if( pxTemplate[ iAttrib ].ulValueLen >= sizeof( CK_OBJECT_CLASS ) )
+                        {
+                            memcpy( pxTemplate[ iAttrib ].pValue, &xClass, sizeof( CK_OBJECT_CLASS ) );
+                        }
+                        else
+                        {
+                            xResult = CKR_BUFFER_TOO_SMALL;
+                        }
+                    }
+
+                    break;
+
+                case CKA_VALUE:
+
+                    if( xIsPrivate == CK_TRUE )
+                    {
+                        pxTemplate[ iAttrib ].ulValueLen = CK_UNAVAILABLE_INFORMATION;
+                        xResult = CKR_ATTRIBUTE_SENSITIVE;
+                    }
+                    else
+                    {
+                        if( pxTemplate[ iAttrib ].pValue == NULL )
+                        {
+                            pxTemplate[ iAttrib ].ulValueLen = ulLength;
+                        }
+                        else if( pxTemplate[ iAttrib ].ulValueLen < ulLength )
+                        {
+         
```

Comment:
this else should probably be a catch all for everything that's not ecdsa

Reply:
Since there's no way to get an RSA key handle out of this implementation in the first place, this would seem to be a fair amount of redundant logic. You may have something more concise in mind, though. Can you please clarify?

Reply:
that's what I had in mind... in a couple places in the code, we have separate error handling for the RSA mechanism which isn't supported from all the other unsupported mechanisms, and I think there is no need for that.  Especially if it is causing different error codes or inconsistent behavior.

Reply:
Agreed - can you please see the latest?